File: scene_graph.py
from dataclasses import dataclass, field
from enum import Enum
from typing import Dict, List, Optional, Tuple, Iterable, Set

# -------- Core types --------

class SceneType(str, Enum):
    ROOT = "root"          # Single entry point
    CHOICE = "choice"      # Pauses gameplay; no video generated
    EXTENSION = "extension"# Continues last video; can have prompt or not
    END = "end"            # Terminal node

SceneId = str

@dataclass
class Edge:
    """Directed edge to another scene. 'label' shows up on CHOICE scenes."""
    target: SceneId
    label: Optional[str] = None  # Required for CHOICE scenes; ignored otherwise

@dataclass
class Scene:
    id: SceneId
    kind: SceneType
    name: Optional[str] = None
    # For CHOICE, outgoing edges must have labels.
    # For EXTENSION/ROOT, you can have a single implicit continuation or branches.
    # For END, there must be no edges.
    edges: List[Edge] = field(default_factory=list)

    # --- EXTENSION-only payload ---
    # You generate at most 8s clips; chain more by setting segments > 1.
    segments: int = 1                    # Number of 8s clips to stitch
    duration_per_segment_sec: int = 8    # Usually fixed at 8
    # Prompt handling:
    # - If prompt is not None, use it verbatim
    # - If prompt is None and inherit_prompt=True, use last choice label
    prompt: Optional[str] = None
    inherit_prompt: bool = True

    # Optional metadata for production
    notes: Optional[str] = None

@dataclass
class SceneGraph:
    """A scene graph indexed by scene id."""
    nodes: Dict[SceneId, Scene]

    def root(self) -> Scene:
        roots = [s for s in self.nodes.values() if s.kind == SceneType.ROOT]
        if len(roots) != 1:
            raise ValueError(f"Graph must have exactly one root, found {len(roots)}.")
        return roots[0]

    # ---- Validation rules based on your spec ----
    def validate(self) -> None:
        r = self.root()

        for s in self.nodes.values():
            if s.kind == SceneType.CHOICE:
                if not s.edges:
                    raise ValueError(f"CHOICE '{s.id}' must have at least one option.")
                if any(e.label is None for e in s.edges):
                    raise ValueError(f"CHOICE '{s.id}' edges must have labels.")
            elif s.kind == SceneType.END:
                if s.edges:
                    raise ValueError(f"END '{s.id}' must not have outgoing edges.")
            # EXTENSION/ROOT can have 0..n edges

        # Basic reachability: all nodes should be reachable from root
        reachable = set(self._dfs_ids(r.id))
        unreachable = set(self.nodes.keys()) - reachable
        if unreachable:
            raise ValueError(f"Unreachable scenes: {sorted(unreachable)}")

    def _dfs_ids(self, start: SceneId) -> Iterable[SceneId]:
        seen: Set[SceneId] = set()
        stack = [start]
        while stack:
            sid = stack.pop()
            if sid in seen: 
                continue
            seen.add(sid)
            for e in self.nodes[sid].edges:
                stack.append(e.target)
        return seen

    # ---- Utilities ----

    def enumerate_paths(self) -> List[List[Tuple[Scene, Optional[str]]]]:
        """
        Returns all root->end paths.
        Each step is (Scene, choice_label_taken_on_incoming_edge).
        """
        start = self.root()
        out: List[List[Tuple[Scene, Optional[str]]]] = []

        def walk(cur: Scene, path: List[Tuple[Scene, Optional[str]]]):
            if cur.kind == SceneType.END or not cur.edges:
                out.append(path + [(cur, None)])
                return
            for e in cur.edges:
                nxt = self.nodes[e.target]
                walk(nxt, path + [(cur, e.label)])

        walk(start, [])
        return out

    def total_render_time_sec(self, path: List[Tuple[Scene, Optional[str]]]) -> int:
        """Sum of all extension (and root-if-extension-like) video time along a path."""
        total = 0
        for scene, _ in path:
            if scene.kind in (SceneType.ROOT, SceneType.EXTENSION):
                total += scene.segments * scene.duration_per_segment_sec
        return total

    def pretty_print(self) -> None:
        """ASCII tree from the unique root."""
        def show(node: Scene, indent: str = ""):
            head = f"{node.kind.value.upper()}: {node.name or node.id}"
            if node.kind in (SceneType.ROOT, SceneType.EXTENSION):
                head += f"  [segments={node.segments} x {node.duration_per_segment_sec}s, " \
                        f"prompt={'\"'+node.prompt+'\"' if node.prompt is not None else ('inherit' if node.inherit_prompt else 'none')}]"
            print(indent + head)
            if not node.edges:
                return
            for i, e in enumerate(node.edges):
                last = (i == len(node.edges) - 1)
                branch = "└─" if last else "├─"
                label = f" ({e.label})" if e.label else ""
                print(indent + f"{branch}→ {e.target}{label}")
                show(self.nodes[e.target], indent + ("   " if last else "│  "))
        show(self.root())

# -------- Example graph matching your scenario --------
# Flow:
# Root -> (ext) Enter dungeon (2 clips) -> Choice (Lockpick/Bash)
#   Lockpick -> (ext) Opening door (1 clip) -> End
#   Bash     -> (ext) Smashing door (1 clip) -> End

graph = SceneGraph(
    nodes={
        "root": Scene(
            id="root",
            kind=SceneType.ROOT,
            name="Game Start",
            segments=2,  # two 8s clips stitched (entering dungeon)
            prompt="entering the dungeon with torchlight flicker",
            inherit_prompt=False,
            edges=[Edge(target="door_choice")]
        ),
        "door_choice": Scene(
            id="door_choice",
            kind=SceneType.CHOICE,
            name="At the locked door",
            edges=[
                Edge(target="lockpick_ext", label="Lockpick door"),
                Edge(target="bash_ext", label="Bash door open"),
            ],
        ),
        "lockpick_ext": Scene(
            id="lockpick_ext",
            kind=SceneType.EXTENSION,
            name="Lockpicking",
            segments=1,                  # one 8s extension
            prompt=None,                 # no explicit prompt → inherit last choice label
            inherit_prompt=True,
            edges=[Edge(target="end")]
        ),
        "bash_ext": Scene(
            id="bash_ext",
            kind=SceneType.EXTENSION,
            name="Bashing",
            segments=1,
            prompt=None,                 # inherit last choice label ("Bash door open")
            inherit_prompt=True,
            edges=[Edge(target="end")]
        ),
        "end": Scene(
            id="end",
            kind=SceneType.END,
            name="The Door Opens"
        ),
    }
)

if __name__ == "__main__":
    graph.validate()
    graph.pretty_print()

    # Get all playable routes and their total render time
    paths = graph.enumerate_paths()
    for i, p in enumerate(paths, 1):
        seconds = graph.total_render_time_sec(p)
        route = " -> ".join(f"{s.id}" + (f"[{lbl}]" if lbl else "") for s, lbl in p)
        print(f"\nRoute {i}: {route}\nTotal video time: {seconds}s")

File: app/lib/sceneGraph.ts
/**
 * @license
 * SPDX-License-Identifier: Apache-2.0
 */

import { log } from './logger';

/**
 * Scene types matching the Python implementation
 */
export enum SceneType {
  ROOT = 'root',
  CHOICE = 'choice',
  EXTENSION = 'extension',
  END = 'end',
}

export type SceneId = string;

/**
 * Directed edge to another scene
 */
export interface Edge {
  target: SceneId;
  label?: string; // Required for CHOICE scenes; ignored otherwise
  prompt?: string; // Prompt for this choice's video (CHOICE scenes only)
}

/**
 * Scene node in the graph
 */
export interface Scene {
  id: SceneId;
  kind: SceneType;
  name?: string;
  edges: Edge[];

  // EXTENSION/ROOT specific properties
  segments?: number; // Number of 8s clips to stitch
  duration_per_segment_sec?: number; // Usually fixed at 8
  prompt?: string | null;
  inherit_prompt?: boolean;

  // Optional metadata
  notes?: string;
}

/**
 * Complete scene graph structure
 */
export interface SceneGraphData {
  nodes: Record<SceneId, Scene>;
}

/**
 * Scene graph with validation and path enumeration
 */
export class SceneGraph {
  nodes: Record<SceneId, Scene>;

  constructor(data: SceneGraphData) {
    this.nodes = data.nodes;
  }

  /**
   * Get the unique root node
   */
  root(): Scene {
    const roots = Object.values(this.nodes).filter(s => s.kind === SceneType.ROOT);
    if (roots.length !== 1) {
      throw new Error(`Graph must have exactly one root, found ${roots.length}.`);
    }
    return roots[0];
  }

  /**
   * Validate graph structure according to specification
   */
  validate(): void {
    log.info('SceneGraph', 'Starting validation', { nodeCount: Object.keys(this.nodes).length });

    try {
      const r = this.root();

      for (const s of Object.values(this.nodes)) {
        if (s.kind === SceneType.CHOICE) {
          if (!s.edges || s.edges.length === 0) {
            throw new Error(`CHOICE '${s.id}' (${s.name || 'unnamed'}) must have at least one option. Click the node, add choice options, and connect child nodes.`);
          }
          const missingLabels = s.edges.filter(e => !e.label).length;
          if (missingLabels > 0) {
            throw new Error(`CHOICE '${s.id}' (${s.name || 'unnamed'}) has ${missingLabels} edge(s) without labels. Edit the node and click 'Save Choice Options'.`);
          }
          const missingPrompts = s.edges.filter(e => !e.prompt).length;
          if (missingPrompts > 0) {
            throw new Error(`CHOICE '${s.id}' (${s.name || 'unnamed'}) has ${missingPrompts} edge(s) without prompts. Edit the node, add prompts, and click 'Save Choice Options'.`);
          }
          // Validate that each choice has exactly one child node
          const childCounts = new Map<string, number>();
          for (const edge of s.edges) {
            childCounts.set(edge.target, (childCounts.get(edge.target) || 0) + 1);
          }
          if (childCounts.size !== s.edges.length) {
            throw new Error(`CHOICE '${s.id}' must have unique children for each choice option.`);
          }
        } else if (s.kind === SceneType.END) {
          if (s.edges && s.edges.length > 0) {
            throw new Error(`END '${s.id}' must not have outgoing edges.`);
          }
        }
        // EXTENSION/ROOT can have 0..n edges
      }

      // Check reachability
      const reachable = new Set(this.dfsIds(r.id));
      const unreachable = Object.keys(this.nodes).filter(id => !reachable.has(id));
      if (unreachable.length > 0) {
        throw new Error(`Unreachable scenes: ${unreachable.join(', ')}`);
      }

      log.info('SceneGraph', 'Validation passed', { reachableNodes: reachable.size });
    } catch (error) {
      log.error('SceneGraph', 'Validation failed', error as Error);
      throw error;
    }
  }

  /**
   * DFS to find all reachable node IDs
   */
  private dfsIds(start: SceneId): Set<SceneId> {
    const seen = new Set<SceneId>();
    const stack = [start];

    while (stack.length > 0) {
      const sid = stack.pop()!;
      if (seen.has(sid)) continue;
      seen.add(sid);

      const node = this.nodes[sid];
      if (node && node.edges) {
        for (const e of node.edges) {
          stack.push(e.target);
        }
      }
    }

    return seen;
  }

  /**
   * Get parent nodes for a given node
   */
  getParents(nodeId: SceneId): Scene[] {
    const parents: Scene[] = [];
    for (const node of Object.values(this.nodes)) {
      if (node.edges?.some(e => e.target === nodeId)) {
        parents.push(node);
      }
    }
    return parents;
  }

  /**
   * Get the edge label from parent to child (for inherit_prompt)
   */
  getIncomingEdgeLabel(nodeId: SceneId): string | null {
    for (const node of Object.values(this.nodes)) {
      if (node.edges) {
        const edge = node.edges.find(e => e.target === nodeId);
        if (edge) {
          return edge.label || null;
        }
      }
    }
    return null;
  }

  /**
   * Enumerate all paths from root to end
   */
  enumeratePaths(): Array<Array<{ scene: Scene; choiceLabel?: string }>> {
    const start = this.root();
    const paths: Array<Array<{ scene: Scene; choiceLabel?: string }>> = [];

    const walk = (
      cur: Scene,
      path: Array<{ scene: Scene; choiceLabel?: string }>
    ): void => {
      if (cur.kind === SceneType.END || !cur.edges || cur.edges.length === 0) {
        paths.push([...path, { scene: cur }]);
        return;
      }

      for (const e of cur.edges) {
        const nxt = this.nodes[e.target];
        if (nxt) {
          walk(nxt, [...path, { scene: cur, choiceLabel: e.label }]);
        }
      }
    };

    walk(start, []);
    return paths;
  }

  /**
   * Calculate total render time for a path
   */
  totalRenderTimeSec(path: Array<{ scene: Scene; choiceLabel?: string }>): number {
    let total = 0;
    for (const { scene } of path) {
      if (scene.kind === SceneType.ROOT || scene.kind === SceneType.EXTENSION) {
        const segments = scene.segments || 1;
        const duration = scene.duration_per_segment_sec || 8;
        total += segments * duration;
      }
    }
    return total;
  }

  /**
   * Convert to JSON-serializable format
   */
  toJSON(): SceneGraphData {
    return { nodes: this.nodes };
  }

  /**
   * Create from JSON data
   */
  static fromJSON(data: SceneGraphData): SceneGraph {
    return new SceneGraph(data);
  }
}

